### PR TITLE
Add BinaryFolder + binary compatible

### PR DESCRIPTION
**Pull request by samuelgarcia**
This PR should remove a deep bug in spikeinterface core.
In some situation using dump to json could lost the propertis (and so the probe).

This should fix this.

It also add a better way to avoid copy for sorter when not needed.

### CODE REVIEW

**Comment by samuelgarcia** [comment]
@alejoe91 : ready to review
tests pass

File: spikeinterface/core/baserecording.py [hunk 1]
```diff
@@ -680,6 +683,59 @@ def planarize(self, axes: str = "xy"):
         recording2d.set_probe(probe2d, in_place=True)
 
         return recording2d
+    
+    def is_binary_compatible(self):
+        """
+        Inform is this recording is "binary" compatible.
```
Comment (alejoe91) [inline spikeinterface/core/baserecording.py:689]:
```suggestion
        Check if the recording is "binary" compatible.
```

File: spikeinterface/core/binaryfolder.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+from .base import _make_paths_absolute
+from .binaryrecordingextractor import BinaryRecordingExtractor
+from .core_tools import define_function_from_class
+
+
+
+class BinaryFolderRecording(BinaryRecordingExtractor):
+    """
+    BinaryFolderRecording is an internal format used in spikeinterface.
+    It is a BinaryRecordingExtractor + metadata contain in a folder.
```
Comment (alejoe91) [inline spikeinterface/core/binaryfolder.py:15]:
```suggestion
    It is a BinaryRecordingExtractor + metadata contained in a folder.
```

File: spikeinterface/core/binaryfolder.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    
+    It is create with any `rec.save(format='binary', folder='/myfolder')`
```
Comment (alejoe91) [inline spikeinterface/core/binaryfolder.py:17]:
```suggestion
    It is created with the function: `recording.save(format='binary', folder='/myfolder')`
```

File: spikeinterface/sorters/kilosortbase.py [hunk 4]
```diff
@@ -113,16 +114,16 @@ def _get_specific_options(cls, ops, params):
...
-            binary_file_path = Path(recording._kwargs['file_paths'][0])
+        if recording.binary_compatible_with(dtype='int16', time_axis=0, file_paths_lenght=1):
+            # no copy
+            d = recording.get_binary_description()
+            binary_file_path = Path(d['file_paths'][0])
         else:
+            # local copy needed
             binary_file_path = output_folder / 'recording.dat'
-            BinaryRecordingExtractor.write_recording(recording, file_paths=binary_file_path,
+            write_binary_recording(recording, file_paths=[binary_file_path],
```
Comment (alejoe91) [inline spikeinterface/sorters/kilosortbase.py:126]: Fix indentation

File: spikeinterface/sorters/klusta/klusta.py [hunk 5]
```diff
@@ -88,21 +88,16 @@ def _setup_recording(cls, recording, output_folder, params, verbose):
...
+            # no copy
+            d = recording.get_binary_description()
+            raw_filename = str(d['file_paths'][0])
+            dtype = str(d['dtype'])
         else:
             # save binary file (chunk by chunk) into a new file
             raw_filename = output_folder / 'recording.dat'
             dtype = 'int16'
-            BinaryRecordingExtractor.write_recording(recording, file_paths=[raw_filename], verbose=False, 
+            write_binary_recording(recording, file_paths=[raw_filename], verbose=False,
```
Comment (alejoe91) [inline spikeinterface/sorters/klusta/klusta.py:100]: fix indentation

File: spikeinterface/sorters/tridesclous/tridesclous.py [hunk 6]
```diff
@@ -87,22 +87,21 @@ def _setup_recording(cls, recording, output_folder, params, verbose):
...
+            file_offset = d['file_offset']
         else:
             if verbose:
                 print('Local copy of recording')
             # save binary file (chunk by chunk) into a new file
             num_chan = recording.get_num_channels()
             dtype = recording.get_dtype().str
             file_paths = [str(output_folder / f'raw_signals_{i}.raw') for i in range(num_seg)]
-            BinaryRecordingExtractor.write_recording(recording, file_paths=file_paths,
+            write_binary_recording(recording, file_paths=file_paths,
```
Comment (alejoe91) [inline spikeinterface/sorters/tridesclous/tridesclous.py:104]: fix indentation

**Comment by samuelgarcia** [comment]
@alejoe91 : done!

**Review by alejoe91 — APPROVED** [review]

**Pull request merged**